### PR TITLE
Eliminate special-casing from test CI workflows.

### PR DESCRIPTION
**Pull request by obi1kenobi**
The previous approach was relying on `_test.yml` taking an input parameter, and then doing almost completely orthogonal things for each parameter value. I've separated out each of those test situations as its own job or workflow file, which eliminated all the special-casing and, in my opinion, improved maintainability by making it much more obvious what code runs when.

### CODE REVIEW

**Comment by obi1kenobi** [comment]
I cherry-picked from #9478 the fix for the places where the pydantic shim wasn't being used but should have been, because the new pydantic compatibility check for the experimental package is failing without it. Yay new tests successfully finding issues!

**Comment by obi1kenobi** [comment]
Hmm it seems like there are more pydantic compatibility issues in the experimental package. I'm going to remove the new check from this commit for the sake of making this easier to merge, and I can add it separately later once we've resolved those issues.

File: .github/workflows/_pydantic_compatibility.yml [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+      - uses: actions/checkout@v3
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: "./.github/actions/poetry_setup"
+        with:
+          python-version: ${{ matrix.python-version }}
+          working-directory: ${{ inputs.working-directory }}
+          poetry-version: ${{ env.POETRY_VERSION }}
+          cache-key: pydantic-cross-compat
+          install-command: poetry install
+      - name: Install the opposite major version of pydantic
```
Comment (eyurtsev) [inline .github/workflows/_pydantic_compatibility.yml:38]: lol clever

**Review by eyurtsev — APPROVED** [review]

File: .github/workflows/_pydantic_compatibility.yml (hunk 1 above)
Comment (obi1kenobi) [reply .github/workflows/_pydantic_compatibility.yml:38]: thank you! 🚀

**Pull request merged**